mautic/mautic
# Fixes API issues with boolean values and applies contact custom field defaults for POST/PUT

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | y
| New feature? | enhancement
| Related user documentation PR URL | na
| Related developer documentation PR URL | na
| Issues addressed (#s or URLs) | na
| BC breaks? | n
| Deprecations? | n

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )


#### Description:
This PR addresses several issues with the API. It is already part of https://github.com/mautic/mautic/pull/2738. 

1. Default values of entity properties was ignored and thus these properties were not included in form validation or new entities. For example, Mautic/EmailBundle/Entity/Email::$emailType. This was resolved by extracting the default values via Reflection to force them as part of the request.
2. Booleans fields did not work when updating an entity due to possibly a bug in Symfony where PATCH ignored false/0 boolean values. A work around was put into place by manually handling boolean fields outside of the form. 
3. For contacts, if the request is POST (new contact) or PUT (updating a contact with a complete representation of the contact) and the request does not include a field with a default value, the default value is populated into the request.

#### Steps to test this PR:
1. Best to test in the context of https://github.com/mautic/mautic/pull/2738 which includes write capability for Emails. Create a new email through the new endpoint leaving out emailType from the request. The default of `template` should be populated.
2. Update a contact using PATCH that includes a boolean field. Can use 1, 0, yes, no, true, false as the values and they should update appropriately. 
3. Create a new contact through the API. Do not include a parameter for a custom field that has a default value defined. The new contact should result in having the default value populated.

[//]: # ( As applicable: )


#### Steps to reproduce the bug:
1. Hard to reproduce without the new API endpoints.
2. Repeat number 2 above with a false value and note that the boolean field will not change from a true value.
